File: tests/cpp/test_route_planner.cpp
#include <algorithm>
#include <cmath>
#include <string>
#include <vector>

#include "qrics/common/types.hpp"
#include "qrics/planning/route_planner.hpp"
#include "qrics/scenario/scene_profile.hpp"

namespace {

[[nodiscard]] qrics::scenario::SceneProfile make_blocked_scene() {
  qrics::scenario::SceneObstacle obstacle{};
  obstacle.obstacle_id = "center_barrel";
  obstacle.geometry_type = qrics::scenario::SceneGeometryType::Cylinder;
  obstacle.pose.position = qrics::common::Vec3{0.55, 0.0, 0.25};
  obstacle.radius_m = 0.14;
  obstacle.height_m = 0.40;

  qrics::scenario::SceneProfile scene{};
  scene.scene_id = "cpp_route_planner_test_scene";
  scene.version = "0.1.0";
  scene.terrain_pack = "mixed_terrain_pack";
  scene.obstacles.push_back(obstacle);
  return scene;
}

[[nodiscard]] double distance_to_obstacle(const qrics::common::Vec3& point) {
  return std::hypot(point.x - 0.55, point.y) - 0.14;
}

[[nodiscard]] bool route_uses_detour(const qrics::planning::PlannedRoute& route) {
  return std::ranges::any_of(route.waypoints, [](const qrics::planning::RouteWaypoint& waypoint) {
    return waypoint.is_detour;
  });
}

}  // namespace

int main() {
  qrics::planning::RoutePlanningConfig config{};
  config.robot_radius_m = 0.22;
  config.safety_margin_m = 0.06;
  config.grid_resolution_m = 0.05;

  const auto route = qrics::planning::plan_task_route(
      make_blocked_scene(),
      std::vector<qrics::planning::RouteTarget>{
          qrics::planning::RouteTarget{"A", qrics::common::Vec3{1.40, 0.0, 0.35}, 0.0}},
      config);
  if (!route.ok) {
    return 1;
  }
  if (route.value.original_target_count != 1 || route.value.blocked_object_count != 1) {
    return 2;
  }
  if (!route.value.used_graph_search || route.value.detour_waypoint_count <= 0 ||
      !route_uses_detour(route.value)) {
    return 3;
  }
  if (route.value.waypoints.empty() || route.value.waypoints.back().waypoint_id != "A") {
    return 4;
  }
  for (const auto& waypoint : route.value.waypoints) {
    if (waypoint.is_detour && std::abs(waypoint.position.y) <= 0.20) {
      return 5;
    }
    if (distance_to_obstacle(waypoint.position) <= config.robot_radius_m) {
      return 6;
    }
  }
  const std::string note_blob = route.value.notes.empty() ? "" : route.value.notes.front();
  if (note_blob.find("C++ route planner") == std::string::npos) {
    return 7;
  }
  return 0;
}

File: tests/python/test_route_planner.py
import math
from dataclasses import dataclass

from qrics.sim import ForbiddenZone, SceneObstacle, SceneProfile, Vec3
from qrics.sim.route_planner import plan_task_route


@dataclass(frozen=True)
class Target:
    target_id: str
    x: float
    y: float
    dwell_steps: int = 0


def _segment_samples(
    start: tuple[float, float], end: tuple[float, float], *, step_m: float = 0.025
) -> list[tuple[float, float]]:
    distance = math.hypot(end[0] - start[0], end[1] - start[1])
    count = max(1, int(math.ceil(distance / step_m)))
    return [
        (
            start[0] + (end[0] - start[0]) * index / count,
            start[1] + (end[1] - start[1]) * index / count,
        )
        for index in range(count + 1)
    ]


def test_route_planner_inserts_via_waypoints_around_obstacle() -> None:
    scene = SceneProfile(
        scene_id="obstacle_route",
        version="0.1.0",
        obstacle_set=(
            SceneObstacle(
                obstacle_id="barrel_on_direct_path",
                position=Vec3(x=0.45, y=0.0, z=0.35),
                radius_m=0.12,
                height_m=0.35,
            ),
        ),
    )

    route = plan_task_route(scene=scene, targets=(Target("A", 0.9, 0.0),))

    assert route.detour_waypoint_count > 0
    assert route.waypoints[-1].target_id == "A"
    assert any(abs(waypoint.y) > 0.20 for waypoint in route.waypoints)

    points = [(0.0, 0.0), *((waypoint.x, waypoint.y) for waypoint in route.waypoints)]
    for start, end in zip(points, points[1:], strict=False):
        for x, y in _segment_samples(start, end):
            clearance = math.hypot(x - 0.45, y - 0.0) - 0.12
            assert clearance > 0.20


def test_route_planner_treats_low_friction_zone_as_blocked() -> None:
    scene = SceneProfile(
        scene_id="no_go_route",
        version="0.1.0",
        forbidden_zones=(
            ForbiddenZone(
                zone_id="low_friction_no_go",
                polygon=(
                    Vec3(0.35, -0.20, 0.0),
                    Vec3(0.65, -0.20, 0.0),
                    Vec3(0.65, 0.20, 0.0),
                    Vec3(0.35, 0.20, 0.0),
                ),
            ),
        ),
    )

    route = plan_task_route(scene=scene, targets=(Target("A", 0.95, 0.0),))

    assert route.detour_waypoint_count > 0
    points = [(0.0, 0.0), *((waypoint.x, waypoint.y) for waypoint in route.waypoints)]
    for start, end in zip(points, points[1:], strict=False):
        for x, y in _segment_samples(start, end):
            assert not (0.25 <= x <= 0.75 and -0.30 <= y <= 0.30)
